### ISSUE TITLE
NLP - Lecture 6

### ISSUE THREAD

**Issue by AranScope** [AranScope/y2-lectures#25]
First up, [here](http://colah.github.io/posts/2014-07-NLP-RNNs-Representations/) is a nice breakdown of the following lecture.

Let's move on to deep learning and word embeddings. We'll start with **linear regression**.

![linear regression graph](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Linear_regression.svg/2000px-Linear_regression.svg.png)

We are looking for a function that best fits a given data set. In the case of the above scatter graph, we can see a linear line, passing through y = 5. The reason this is useful, is because it allows us to identify patterns in data.
## Least square linear regression

With least square we are trying to minimise the total absolute distance in Y between the line and each of the data points. 

-- differentiation --
### Linear regression in higher dimensions

equation of a line - `y = mx + b` using matrices - `Y = mX + b`
#### Gradient descent

Differentiation is a computationally expensive task, using an algorithm such as **gradient descent** will increase the efficient of the linear regression algorithm.
## Logistic regression

Logistic regression is used for categorisation, think classifying words, determining the language of text etc. See **non-linear support vector machine**.
## Neural networks

Neutral networks have several logistic regression units.

Some fun bedtime reading can be found [Here!](https://en.wikipedia.org/wiki/Convolutional_neural_network)
#### Deep learning in NLP
- Next work prediction problem
  - Predictive text
- Grammar check problem
#### Word embeddings

`'cat','sat','on','the','mat'` vs `'cat','sat','song','the','mat'`. For the first sentence we want the neural network to return '1' or, some confirmation that the sentence is 'valid'. For the second, 0. 

[Here](http://stevenmiller888.github.io/mind-how-to-build-a-neural-network/) is some more fun bedtime reading.

We can train a neural network with a large dataset, after each datum is entered we modify the variables in the neural network to ensure that the output is consistently correct.
#### Shared representations

Take some text and it's corresponding text in say, a different language. We will form a relation between these stating that they are similar. This is useful for machine translation.
### Word2vec

Pre-trained word and phrase vectors, released by Google. The dataset is ~ 100B words and contains 300 dimensional vectors for 3 million words and phrases.

Phrases are seperated because words; when combined, can have completely different meaning to their constituent parts. e.g. `sledge vs. sledge hammer`.
### Fasttest

**It's fast!**, Facebook's fasttext can be trained in the order of seconds vs. hours, days or weeks taken by its competitors.
